antlr_plsql/ast.py

```diff
@@ -313,14 +352,254 @@ class CaseWhen(AstNode):
+                columns_ctx, predicate=is_terminal, simplify=False
+            )
+
+            reference_ctx = foreign_key_ctx.references_clause()
+            obj.reference = visitor.visit(reference_ctx)
+        elif ctx.CHECK():
+            obj.type = "check"
+        return obj
+
+
+class InsertStmt(AstNode):
+    _rules = [("insert_statement", "_from_single_table")]
+    _fields_spec = ["table", "columns", "values", "query"]
+
+    @classmethod
+    def _from_single_table(cls, visitor, ctx):
+        obj = cls._from_fields(visitor, ctx)
+        ctx = ctx.single_table_insert()
+
+        insert_into_ctx = ctx.insert_into_clause()
+        obj.table = visitor.visit(insert_into_ctx.general_table_ref())
+
+        columns_ctx = insert_into_ctx.column_name_list()
+        if columns_ctx:
+            obj.columns = visitor.visit(columns_ctx)
+
+        values_ctx = ctx.values_clause()
+        if values_ctx:
+            values_list_ctx = values_ctx.expression_list()
+            obj.values = visitor.visit(values_list_ctx)
+
+        query_ctx = ctx.select_statement()
+        if query_ctx:
+            obj.query = visitor.visit(query_ctx)
+
+        return obj
+
+
+class UpdateStmt(AstNode):
+    _rules = [("update_statement", "_from_update")]
+    _fields_spec = [
+        "general_table_ref->table",
+        "where_clause",
+        "from_clause",
+        "updates",
+    ]
+
+    @classmethod
+    def _from_update(cls, visitor, ctx):
+        obj = cls._from_fields(visitor, ctx)
+
+        # TODO extract as visitor ~ expression_list
+        update_set_ctx = ctx.update_set_clause()
+        if update_set_ctx.column_based_update_set_clause():
+            obj.updates = visitor.visitChildren(
+                update_set_ctx, predicate=is_terminal, simplify=False
+            )
+
+        return obj
+
+
+class Update(AstNode):
+    # TODO: BinExpr? Not fit for multiple columns combined?
+    _rules = ["column_based_update_set_clause"]
+    _fields_spec = ["column_name->column", "expression"]
+
+
+class DeleteStmt(AstNode):
+    _rules = ["delete_statement"]
+    _fields_spec = ["general_table_ref->table", "where_clause"]
+
+
 # class FunctionArgument
 
 
+class Terminal(AstNode):
+    _fields_spec = ["value"]
+    DEBUG_INSTANCES = []
+
+    def __new__(cls, *args, **kwargs):
+        instance = super().__new__(cls, *args, **kwargs)
+        if DEBUG:
+            cls.DEBUG_INSTANCES.append(instance)
+            return instance
+        else:
+            return kwargs.get("value", "")
+
+    def __str__(self):
+        return self.value
+
+
 # PARSE TREE VISITOR ----------------------------------------------------------
 
 
 class AstVisitor(grammar.Visitor):
-    def visitChildren(self, node, predicate=None):
+    def visitChildren(self, node, predicate=None, simplify=True):
+        """This is the default visiting behaviour
+
+        :param node: current node
+        :param predicate: skip a child if this evaluates to false
```

Comment:
By using the name `predicate`, it makes me think we're only going to visit children for which predicate is `True`. It's the other way around. Maybe `skip` or `skip_when`.

Reply:
Also, you don't check whether it's `callable` below.